src/cpp/fastdds/xtypes/dynamic_types/DynamicTypeBuilderImpl.cpp

```diff
@@ -0,0 +1,639 @@
+
+    if (member_by_name_.end() == it)
+    {
+        EPROSIMA_LOG_ERROR(DYN_TYPES, "Cannot find a member with name " << name.c_str());
+        return RETCODE_BAD_PARAMETER;
+    }
+
+    member = it->second;
+    return RETCODE_OK;
+}
+
+ReturnCode_t DynamicTypeBuilderImpl::get_all_members_by_name(
+        DynamicTypeMembersByName& member) noexcept
+{
+    member = member_by_name_;
+    return RETCODE_OK;
+}
+
+ReturnCode_t DynamicTypeBuilderImpl::get_member(
+        traits<DynamicTypeMember>::ref_type& member,
+        MemberId id) noexcept
+{
+    auto it = member_.find(id);
+
+    if (member_.end() == it)
+    {
+        EPROSIMA_LOG_ERROR(DYN_TYPES, "Cannot find a member with MemberId " << id);
+        return RETCODE_BAD_PARAMETER;
+    }
+
+    member = it->second;
+    return RETCODE_OK;
+}
+
+ReturnCode_t DynamicTypeBuilderImpl::get_all_members(
+        DynamicTypeMembersById& member) noexcept
+{
+    member = member_;
+    return RETCODE_OK;
+}
+
+uint32_t DynamicTypeBuilderImpl::get_member_count() noexcept
+{
+    return static_cast<uint32_t>(members_.size());
+}
+
+ReturnCode_t DynamicTypeBuilderImpl::get_member_by_index(
+        traits<DynamicTypeMember>::ref_type& member,
+        uint32_t index) noexcept
+{
+    if (index >= members_.size())
+    {
+        EPROSIMA_LOG_ERROR(DYN_TYPES, "Index " << index << " out-of-range");
+        return RETCODE_BAD_PARAMETER;
+    }
+
+    member = members_.at(index);
+    return RETCODE_OK;
+}
+
+uint32_t DynamicTypeBuilderImpl::get_annotation_count() noexcept
+{
+    return static_cast<uint32_t>(annotation_.size());
+}
+
+ReturnCode_t DynamicTypeBuilderImpl::get_annotation(
+        traits<AnnotationDescriptor>::ref_type& descriptor,
+        uint32_t idx) noexcept
+{
+    if (!descriptor || idx >= annotation_.size())
+    {
+        EPROSIMA_LOG_ERROR(DYN_TYPES, "Descriptor reference is nil or index is out-of-range");
+        return RETCODE_BAD_PARAMETER;
+    }
+
+    traits<AnnotationDescriptor>::narrow<AnnotationDescriptorImpl>(descriptor)->copy_from(annotation_.at(idx));
+    return RETCODE_OK;
+}
+
+bool DynamicTypeBuilderImpl::equals(
+        traits<DynamicType>::ref_type other) noexcept
+{
+    bool ret_value = true;
+    auto impl = traits<DynamicType>::narrow<DynamicTypeImpl>(other);
+
+    ret_value &= type_descriptor_.equals(impl->type_descriptor_);
+    if (ret_value)
+    {
+        ret_value &= annotation_.size() == impl->annotation_.size();
+        if (ret_value)
+        {
+            for (size_t count {0}; ret_value && count < annotation_.size(); ++count)
+            {
+                ret_value &= annotation_.at(count).equals(impl->annotation_.at(count));
+            }
+        }
+
+        ret_value &= member_.size() == impl->member_.size();
+        assert(TK_ANNOTATION == type_descriptor_.kind() ||
+                TK_STRUCTURE == type_descriptor_.kind() ||
+                TK_UNION ==  type_descriptor_.kind() ||
+                0 == member_.size());
```

Comment:
DynamicTypeImpl::equals takes into account that bitmask and bitset also have members. But DynamicTypeBuilderImpl does not take them into account. Also, DynamicTypeBuilderImpl::add_member allows to add members to enum types. I think there are inconsistencies in the implementation that need to be fixed.

Reply:
Done in c8b1745ba